internal/git/util.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/numaproj-labs/numaplane/internal/kustomize"
+	"github.com/numaproj-labs/numaplane/internal/shared/logging"
+	k8sClient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	controllerConfig "github.com/numaproj-labs/numaplane/internal/controller/config"
+	gitShared "github.com/numaproj-labs/numaplane/internal/shared/git"
+)
+
+func CloneRepo(
+	ctx context.Context,
+	client k8sClient.Client,
+	gitSync *v1alpha1.GitSync,
+	namespace string,
+	globalConfig controllerConfig.GlobalConfig,
+) (*git.Repository, error) {
+	gitCredentials := gitShared.FindCredByUrl(gitSync.Spec.RepoUrl, globalConfig)
+	cloneOptions, err := gitShared.GetRepoCloneOptions(ctx, gitCredentials, client, namespace, gitSync.Spec.RepoUrl)
+	if err != nil {
+		return nil, fmt.Errorf("error getting  the  clone options: %v", err)
+	}
+	return cloneRepo(ctx, gitSync, cloneOptions)
+}
+
+// GetLatestManifests monitors a Git repository for changes. It fetches updates from the remote repository,
+// checks the latest commit hash against the stored hash in the GitSync object,
+// and applies any changes to the Kubernetes cluster. It also periodically checks for updates based on a ticker.
+func GetLatestManifests(
+	ctx context.Context,
+	r *git.Repository,
+	gitSync *v1alpha1.GitSync,
+
+) ([]string, error) {
+	logger := logging.FromContext(ctx).With("GitSync name", gitSync.Name, "repo", gitSync.Spec.RepoUrl)
+	manifests := make([]string, 0)
+
+	// Fetch all remote branches
+	err := fetchUpdates(r)
+	if err != nil {
+		return manifests, err
+	}
+
+	// The revision can be a branch, a tag, or a commit hash
+	hash, err := getLatestCommitHash(r, gitSync.Spec.TargetRevision)
+	if err != nil {
+		logger.Errorw("error resolving the revision", "revision", gitSync.Spec.TargetRevision, "err", err, "repo", gitSync.Spec.RepoUrl)
+		return manifests, err
+	}
+
+	// Only create the resources for the first time if not created yet.
+	// Otherwise, monitoring with intervals.
+
+	// kustomizePath will be the path of clone repository + the path where kustomize file is present.
+	localRepoPath := getLocalRepoPath(gitSync.GetName())
+	kustomizePath := localRepoPath + "/" + gitSync.Spec.Path
+	k := kustomize.NewKustomizeApp(kustomizePath, gitSync.Spec.RepoUrl, os.Getenv("KUSTOMIZE_BINARY_PATH"))
+
+	if kustomize.IsKustomizationRepository(kustomizePath) {
+		m, err := k.Build(nil)
+		if err != nil {
+			logger.Errorw("cannot build kustomize yaml", "err", err)
+			return manifests, err
+		}
+		manifests = append(manifests, m...)
+	} else {
+		// Retrieving the commit object matching the hash.
+		tree, err := getCommitTreeAtPath(r, gitSync.Spec.Path, *hash)
+		if err != nil {
+			return manifests, err
+		}
+		// Read all the files under the path and apply each one respectively.
+		err = tree.Files().ForEach(func(f *object.File) error {
+			logger.Debugw("read file", "file_name", f.Name)
+			// TODO: this currently assumes that one file contains just one manifest - modify for multiple
```

Comment:
so, this is still not implemented, right? should there be an issue for it?

Reply:
Sure, let me create an issue to track verify if it is still applicable then fix if so.

Reply:
https://github.com/numaproj-labs/numaplane/issues/128

Reply:
thanks!